libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -154,8 +154,14 @@ void AP_NMEA_Output::update()
     char rmc_end[6];
     snprintf(rmc_end, sizeof(rmc_end), "*%02X\r\n", (unsigned) _nmea_checksum(rmc));
 
+    const uint32_t space_required = strlen(gga) + strlen(gga_end) + strlen(rmc) + strlen(rmc_end);
```

Comment:
You have to check that `gga_res` and `rmc_res` are different than -1. If `asnprint` fails, it is undefined what the contents of the pointer are.

Also if `gga` or `rmc` aren't good, we don't send the endings too - so we won't be needing that UART space.

Reply:
Is there some reason we don't early-exit if those ansprintfs fail?  Those failing is probably a sign to start to tread delicately.

Reply:
Well it isn't really defined what errors asnprintf can have, all we know is that it returns -1 on an error, any error - so not necessarily a memory allocation issue.

The reason behind the current code is simply that it was mostly based on the same from SITL GPS. I don't oppose changing it to an early exit, but current code is more functional - if you have at least one message, send it. We could/should early-exit if they are both bad though.

Reply:
OK, I've added early exits.  I don't think "best effort" do get any message out is appropriate here.  

If we wait until we can fit both then a client will get both messages.  If we were to continue if the second asprintf fails then we may starve the client of RMC messages, which it may require.